user/lmbench3/doc/lat_unix.8
.\" $Id: //WIFI_SOC/release/SDK_4_1_0_0/source/user/lmbench3/doc/lat_unix.8#1 $
.TH LAT_UNIX 8 "$Date: 2012/09/28 $" "(c)1994-2000 Carl Staelin and Larry McVoy" "LMBENCH"
.SH NAME
lat_unix \- measure interprocess communication latency via UNIX sockets
.SH SYNOPSIS
.B lat_unix
[
.I "-m <message size>"
]
[
.I "-P <parallelism>"
]
[
.I "-W <warmups>"
]
[
.I "-N <repetitions>"
]
.SH DESCRIPTION
.B lat_unix
is a client/server program that measures interprocess
communication latencies.  The benchmark passes a message back and forth between
the two processes (this sort of benchmark is frequently referred to as a
``hot potato'' benchmark).  No other work is done in the processes.
.SH OUTPUT
The reported time is in microseconds per round trip and includes the total
time, i.e., the context switching overhead is includeded.
Output format is like so
.sp
.ft CB
AF_UNIX sock stream latency: 700 microseconds
.ft
.SH ACKNOWLEDGEMENT
Funding for the development of
this tool was provided by Sun Microsystems Computer Corporation.
.SH "SEE ALSO"
lmbench(8), lat_fcntl(8), lat_fifo(8), lat_tcp(8), lat_udp(8).
.SH "AUTHOR"
Carl Staelin and Larry McVoy
.PP
Comments, suggestions, and bug reports are always welcome.


History
Ralink AP SDK - History

Version 4.1.2.0
===============

BUG FIXED:
- MT7620 USB: Increase MT7620 BUS clock in CPU sleep mode for USB function
- MT7620 PCIe: MT7620 couldn't identify some NIC when it does power on or software reboot
- MT7620 Ethernet: MT7620A WIFI to 2 FE low throughputs
- MT7620 HNAT: MT7620 HNAT accounting and mis-binding isssue
- SPI flash: Flash programming issue when some sector are in write protect mode
- NAND flash: fix bad block offset from byte 5 to byte 0 of spare area for 2k page size case
- System: after GUI upgrades firmware, network disconnects
- GUI: security key checking, WPS Cancel, web server crash, wrong MBSSID counting, WAPI for 2nd wifi

IMPROVEMENT:
- MT7620 Uboot: Increase MT7620 SPI clock for relocating


Version 4.1.1.0
===============

IMPROVEMENT:
- MT7620 Uboot: New add  DDR1 16MB setting: Winbond(W9412G6JH-5) and ESMT(M13S128168A-5TG2N) are already verified.
- MT7620 Uboot: Update DDR2 setting to safer setting. No performance impact.

BUG FIXED:
- MT7620 Linux: Fixed the system timer kernel functions(udelay/HZ) not accurate issue.
- MT7620 HWNAT: not accelerate UDP traffic to fix UDP check=0 issue.


Version 4.1.0.0
===============

PLATFORM:

- support MT7620 (linux-2.6.36.x)
- support RT6855/RT6856 (linux-2.6.36MT.x)
- support RT3052/RT3883/RT3352/RT5350 (linux-2.6.21.x)

NEW FUNCTIONS:

- [Eth] [RT6856/MT7620] support TSO/LRO
- [GUI] Hotspot GUI support
- [GUI] 11AC GUI support
- [UPnP] miniupnpd replaces upnpd and wscd

IMPROVEMENT:

- [HWNAT] change LAN/WAN VID by hw_nat command                                  
- [SW] RT6856 switch table support MLD snooping 
- [PPTP/L2TP] performance enhancement
- [PCIe] Support PCIe Device on PCIe slot 1 but PCIe slot 0 floating 

BUG FIXED:

- [Eth] CDMA checksum offload fail for IP fragment packets                    
- [Busybox] first command failed after Ctrl-C
- [IPv6] IPv6 Multicast Streaming data corrupts on WiFI interface
- [I2S] Add DMA stop handling for last DMA 
- [PCM] [RT6856] Add SPI CTRL for SLIC example





user/lmbench3/src/lat_unix.c
/*
 * tcp_unix.c - simple UNIX socket transaction latency test
 *
 *	lat_unix [-P <parallelism>] [-W <warmup>] [-N <repetitions>]
 *
 * Copyright (c) 1994-2000 Carl Staelin and Larry McVoy.  
 * Distributed under the FSF GPL with additional restriction that 
 * results may published only if
 * (1) the benchmark is unmodified, and
 * (2) the version in the sccsid below is included in the report.
 * Support for this development by Sun Microsystems is gratefully acknowledged.
 */
char	*id = "$Id: //WIFI_SOC/release/SDK_4_1_0_0/source/user/lmbench3/src/lat_unix.c#1 $\n";
#include "bench.h"

struct _state {
	int	sv[2];
	int	pid;
	int	msize;
	char*	buf;
};
void	initialize(iter_t iterations, void* cookie);
void	benchmark(iter_t iterations, void* cookie);
void	cleanup(iter_t iterations, void* cookie);

int
main(int ac, char **av)
{
	int parallel = 1;
	int warmup = 0;
	int repetitions = TRIES;
	struct _state state;
	int c;
	char* usage = "[-m <message size>] [-P <parallelism>] [-W <warmup>] [-N <repetitions>]\n";

	state.msize = 1;
	state.pid = 0;

	while (( c = getopt(ac, av, "m:P:W:N:")) != EOF) {
		switch(c) {
		case 'm':
			state.msize = atoi(optarg);
			break;
		case 'P':
			parallel = atoi(optarg);
			if (parallel <= 0) lmbench_usage(ac, av, usage);
			break;
		case 'W':
			warmup = atoi(optarg);
			break;
		case 'N':
			repetitions = atoi(optarg);
			break;
		default:
			lmbench_usage(ac, av, usage);
			break;
		}
	}
	if (optind < ac) {
		lmbench_usage(ac, av, usage);
	}

	benchmp(initialize, benchmark, cleanup, 0, parallel, 
		warmup, repetitions, &state);

	micro("AF_UNIX sock stream latency", get_n());
	return(0);
}

void
initialize(iter_t iterations, void* cookie)
{
	struct _state* pState = (struct _state*)cookie;
	void	exit();

	if (iterations) return;

	if (socketpair(AF_UNIX, SOCK_STREAM, 0, pState->sv) == -1) {
		perror("socketpair");
	}

	pState->buf = malloc(pState->msize);
	if (pState->buf == NULL) {
		fprintf(stderr, "buffer allocation\n");
		exit(1);
	}
	handle_scheduler(benchmp_childid(), 0, 1);

	if (pState->pid = fork())
		return;

	handle_scheduler(benchmp_childid(), 1, 1);

	/* Child sits and ping-pongs packets back to parent */
	signal(SIGTERM, exit);
	while (read(pState->sv[0], pState->buf, pState->msize) == pState->msize) {
		write(pState->sv[0], pState->buf, pState->msize);
	}
	exit(0);
}

void
benchmark(iter_t iterations, void* cookie)
{
	struct _state* pState = (struct _state*)cookie;

	while (iterations-- > 0) {
		if (write(pState->sv[1], pState->buf, pState->msize) != pState->msize
		    || read(pState->sv[1], pState->buf, pState->msize) != pState->msize) {
			/* error handling: how do we signal failure? */
			cleanup(0, cookie);
			exit(0);
		}
	}
}

void
cleanup(iter_t iterations, void* cookie)
{
	struct _state* pState = (struct _state*)cookie;

	if (iterations) return;

	if (pState->pid) {
		kill(pState->pid, SIGKILL);
		waitpid(pState->pid, NULL, 0);
		pState->pid = 0;
	}
}

